model/src/main/proto/exploration_checkpoint.proto

```diff
@@ -0,0 +1,75 @@
+syntax = "proto3";
+
+package model;
+
+import "exploration.proto";
+
+option java_package = "org.oppia.android.app.model";
+option java_multiple_files = true;
+
+// Top-level proto used to store exploration checkpoint per-profile basis.
+message ExplorationCheckpointDatabase {
+
+  // map from explorationId to ExplorationCheckpoint.
+  map<string, ExplorationCheckpoint> exploration_checkpoint = 1;
+}
+
+// Corresponds to the exploration checkpoint which will be used to resume the exploration.
+// This model contains information about all states of the deck which have been visited by the user.
+message ExplorationCheckpoint {
+
+  // The name of the latest state in the deck.
+  string pending_state_name = 1;
+
+  // The list of all states which the learner has previously completed.
+  repeated CompletedStateInCheckpoint completed_states_in_checkpoint = 2;
+
+  // Stores answers to the latest state in the deck.
+  repeated AnswerAndResponse pending_user_answers = 3;
+
+  // The Index of the current selected state from the deck of states.
+  int32 state_index = 4;
+
+  // Stores -1 if no hint for the pending top state has been revealed otherwise it stores the index
+  // of the hint that the learner has revealed.
+  int32 hint_index = 5;
+
+  // Stores true if the solution to the pending top state has been revealed by the learner otherwise
+  // it stores false.
+  bool solution_is_revealed = 6;
+
+  // The timestamp in milliseconds of when the checkpoint was saved for the first time.
+  int64 timestamp_of_first_checkpoint = 7;
+
+  // Stores the title of the saved exploration.
+  string exploration_title = 8;
+
+  // Stores the version of the exploration which is compatible with the saved checkpoint.
+  int32 exploration_version = 9;
+}
+
+// Corresponds to exploration states that have been completed.
+message CompletedStateInCheckpoint {
+
+  // Corresponds to an exploration state that the learner has previous completed. Contains a list
+  // of AnswerAndResponse for a previous state.
+  CompletedState completed_state = 1;
+
+  // The name of the state previously completed.
+  string state_name = 2;
```

Comment:
Why do we need this state_name ?

Reply:
state_name is used to make the implementation slightly simpler because the current codebase has a function that retrieves a state using the state's name.

Should I remove this?

Reply:
Okay.
I suggest instead of saving `state_name` here when you are fetching the information from database to send it to app layer in that function you can use multiple data providers to fetch the information of checkpoint as well as state_name from two different database and then combine it and then send it to app layer.

@BenHenning Any thoughts on this?

Reply:
Resolving this as it is correct.